File: ros_ws/src/analysis/analysis/scan_to_range.py
#!/usr/bin/env python3
"""Convert a LaserScan from /sonar into a Range message."""

import math

from rclpy import init, shutdown, spin
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import LaserScan, Range


class ScanToRange(Node):
    """Publish a range message using the closest valid beam."""

    def __init__(self) -> None:
        super().__init__("scan_to_range")
        self.declare_parameter("scan_topic", "/sonar")
        self.declare_parameter("range_topic", "/range")

        scan_topic = self.get_parameter("scan_topic").value
        range_topic = self.get_parameter("range_topic").value

        self.publisher = self.create_publisher(Range, range_topic, qos_profile_sensor_data)
        self.create_subscription(
            LaserScan,
            scan_topic,
            self.scan_callback,
            qos_profile_sensor_data,
        )

        self.get_logger().info(f"ScanToRange: {scan_topic} -> {range_topic}")
        # self.get_logger().info("ScanToRange: %s -> %s", scan_topic, range_topic)

    def scan_callback(self, scan: LaserScan) -> None:
        best = None
        for value in scan.ranges:
            if not math.isfinite(value):
                continue
            if scan.range_min <= value <= scan.range_max:
                if best is None or value < best:
                    best = value

        message = Range()
        message.header = scan.header
        message.radiation_type = Range.ULTRASOUND
        message.field_of_view = scan.angle_max - scan.angle_min
        message.min_range = scan.range_min
        message.max_range = scan.range_max
        message.range = best if best is not None else scan.range_max

        if message.range < message.min_range:
            message.range = message.min_range
        if message.range > message.max_range:
            message.range = message.max_range

        self.publisher.publish(message)


def main() -> None:
    init(args=None)
    node = ScanToRange()
    try:
        spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        shutdown()


if __name__ == "__main__":
    main()

File: ros_ws/src/robot_io_adapters/src/scan_to_range.cpp
#include "robot_io_adapters/scan_to_range.hpp"

#include <cmath>

namespace robot_io_adapters
{

double select_best_range(const sensor_msgs::msg::LaserScan & scan)
{
  double best = scan.range_max;

  for (const float value : scan.ranges) {
    if (!std::isfinite(value)) {
      continue;
    }
    if (value < scan.range_min || value > scan.range_max) {
      continue;
    }
    if (value < best) {
      best = value;
    }
  }

  if (best < scan.range_min) {
    return scan.range_min;
  }
  if (best > scan.range_max) {
    return scan.range_max;
  }
  return best;
}

sensor_msgs::msg::Range convert_scan_to_range(const sensor_msgs::msg::LaserScan & scan)
{
  sensor_msgs::msg::Range message;
  message.header = scan.header;
  message.radiation_type = sensor_msgs::msg::Range::ULTRASOUND;
  message.field_of_view = scan.angle_max - scan.angle_min;
  message.min_range = scan.range_min;
  message.max_range = scan.range_max;
  message.range = select_best_range(scan);
  return message;
}

ScanToRangeNode::ScanToRangeNode()
: rclcpp::Node("scan_to_range")
{
  _scan_topic = this->declare_parameter<std::string>("scan_topic", "/sonar");
  _range_topic = this->declare_parameter<std::string>("range_topic", "/range");

  const auto qos = rclcpp::SensorDataQoS();
  _publisher = this->create_publisher<sensor_msgs::msg::Range>(_range_topic, qos);
  _subscription = this->create_subscription<sensor_msgs::msg::LaserScan>(
    _scan_topic,
    qos,
    [this](const sensor_msgs::msg::LaserScan & scan) {scan_callback(scan);});

  RCLCPP_INFO(this->get_logger(), "ScanToRange: %s -> %s", _scan_topic.c_str(),
    _range_topic.c_str());
}

void ScanToRangeNode::scan_callback(const sensor_msgs::msg::LaserScan & scan)
{
  _publisher->publish(convert_scan_to_range(scan));
}

} // namespace robot_io_adapters
